csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -49,62 +51,82 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
     }
   }
 
-  // Fusion outputs that are (1) aliased by another fusion output, (2) not
-  // aliases themselves, and (3) not fusion inputs (yes, a fusion may trivially
-  // forward an input). Code will later add `segment_set` before them so aliases
+  std::vector<Val*> non_alias_outs;
+  non_alias_outs.reserve(fusion->outputs().size());
+  // Fusion inputs/outputs that are (1) aliased by another output and (2) not
+  // an alias itself. Code will later add `segment_set` around them so aliases
   // are separated from non-aliases and more likely to be accepted by the no-op
-  // scheduler.
-  std::unordered_set<TensorView*> aliased_outs;
+  // scheduler. See AliasTest.OutputAliasesAnotherOutput and
+  // AliasTest.SegmentMetaOps.
+  //
+  // This algorithm is suboptimal in many cases. See
+  // https://github.com/NVIDIA/Fuser/issues/2395#issuecomment-2207043749 for a
+  // proposal that may work better.
+  std::unordered_set<TensorView*> aliased_ios;
   for (TensorView* out_tv :
        ir_utils::filterByType<TensorView>(fusion->outputs())) {
     if (TensorView* aliased_io = analysis.getNearestAliasedIo(out_tv)) {
-      if (aliased_io->isFusionOutput() && !aliased_io->isFusionInput() &&
-          analysis.getNearestAliasedIo(aliased_io) == nullptr) {
-        aliased_outs.insert(aliased_io);
+      if (analysis.getNearestAliasedIo(aliased_io) == nullptr) {
+        aliased_ios.insert(aliased_io);
       }
+    } else {
+      non_alias_outs.push_back(out_tv);
     }
   }
 
-  for (TensorView* aliased_out : aliased_outs) {
-    // Rarely, if `aliased_out` is already defined by `segment_set`, don't
-    // create another `segment_set`.
-    if (LoadStoreOp* def =
-            dynamic_cast<LoadStoreOp*>(aliased_out->definition())) {
-      if (def != nullptr && def->opType() == LoadStoreOpType::SegmenterSet) {
+  // Mark all expressions that are (transitively) used by non-alias outputs.
+  auto used_by_non_aliases =
+      [](const std::vector<Expr*>& exprs) -> std::unordered_set<Expr*> {
+    return {exprs.begin(), exprs.end()};
+  }(StmtSort::getExprsTo(non_alias_outs));
+
+  for (TensorView* aliased_io : aliased_ios) {
+    // Divide the users of aliased_io into two groups according to
+    // `used_by_non_aliases`.
+    std::vector<Expr*> users_used_by_non_aliases;
+    std::vector<Expr*> users_used_only_by_aliases;
+    for (Expr* e : aliased_io->uses()) {
+      if (used_by_non_aliases.count(e)) {
+        users_used_by_non_aliases.push_back(e);
+      } else {
+        users_used_only_by_aliases.push_back(e);
+      }
+    }
+
+    // If all users are used by non-alias outputs, do nothing. Adding
+    // segment_set around unlikely creates more no-op regions.
+    if (users_used_only_by_aliases.empty()) {
+      continue;
+    }
+
+    // If all users are used by aliases, put a `segment_set` before it.
+    if (users_used_by_non_aliases.empty()) {
+      if (aliased_io->isFusionInput()) {
```

Comment:
I think this one short-cut should be lifted.

Reply:
I think https://github.com/NVIDIA/Fuser/pull/2529/files#r1669416615 answers this as well.